src/utils.rs

```diff
@@ -238,3 +238,131 @@ where
         Array::from_nested(self)
     }
 }
+
+#[derive(Debug)]
+pub(crate) struct Closure<'a> {
+    c_closure: mlx_closure,
+    lt_marker: PhantomData<&'a ()>,
+}
+
+impl<'a> Closure<'a> {
+    pub(crate) fn from_ptr(c_closure: mlx_closure) -> Self {
+        Self {
+            c_closure,
+            lt_marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn new<F>(closure: F) -> Self
+    where
+        F: FnMut(&[Array]) -> Vec<Array> + 'a,
+    {
+        let c_closure = new_mlx_closure(closure);
+        Self {
+            c_closure,
+            lt_marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn as_ptr(&self) -> mlx_closure {
+        self.c_closure
+    }
+}
+
+impl<'a> Drop for Closure<'a> {
+    fn drop(&mut self) {
+        unsafe { mlx_sys::mlx_free(self.c_closure as *mut _) }
+    }
+}
+
+/// Helper method to create a mlx_closure from a Rust closure.
+fn new_mlx_closure<'a, F>(closure: F) -> mlx_sys::mlx_closure
+where
+    F: FnMut(&[Array]) -> Vec<Array> + 'a,
+{
+    // Box the closure to keep it on the heap
+    let boxed = Box::new(closure);
+
+    // Create a raw pointer from the Box, transferring ownership to C
+    let raw = Box::into_raw(boxed);
+    let payload = raw as *mut std::ffi::c_void;
+
+    unsafe {
+        mlx_sys::mlx_closure_new_with_payload(Some(trampoline::<F>), payload, Some(noop_dtor))
+    }
+}
+
+/// Function to create a new (+1 reference) mlx_vector_array from a vector of Array
+fn new_mlx_vector_array(arrays: Vec<Array>) -> mlx_sys::mlx_vector_array {
+    unsafe {
+        let result = mlx_sys::mlx_vector_array_new();
+        let ctx_ptrs: Vec<mlx_sys::mlx_array> = arrays.iter().map(|array| array.as_ptr()).collect();
+        mlx_sys::mlx_vector_array_add_arrays(result, ctx_ptrs.as_ptr(), arrays.len());
+        result
+    }
+}
+
+fn mlx_vector_array_values(vector_array: mlx_sys::mlx_vector_array) -> Vec<Array> {
+    unsafe {
+        let size = mlx_sys::mlx_vector_array_size(vector_array);
+        (0..size)
+            .map(|index| {
+                // ctx is a +1 reference, the array takes ownership
+                let c_array = mlx_sys::mlx_vector_array_get(vector_array, index);
+                Array::from_ptr(c_array)
+            })
+            .collect()
+    }
+}
+
+extern "C" fn trampoline<'a, F>(
+    vector_array: mlx_sys::mlx_vector_array,
+    payload: *mut std::ffi::c_void,
+) -> mlx_sys::mlx_vector_array
+where
+    F: FnMut(&[Array]) -> Vec<Array> + 'a,
+{
+    unsafe {
+        let raw_closure: *mut F = payload as *mut _;
+        // Let the box take care of freeing the closure
+        let mut closure = Box::from_raw(raw_closure);
+        let arrays = mlx_vector_array_values(vector_array);
+        let result = closure(&arrays);
+        // We should probably keep using new_mlx_vector_array here instead of VectorArray
+        // since we probably don't want to drop the arrays in the closure
+        new_mlx_vector_array(result)
```

Comment:
why would things get dropped if we use VectorArray?

Reply:
Because we have implemented Drop for VectorArray right?